mods/bucket/init.lua

```diff
@@ -153,6 +153,22 @@ minetest.register_craftitem("bucket:bucket_empty", {
 			minetest.add_node(pointed_thing.under, {name="air"})
 
 			return ItemStack(giving_back)
+		else
+			local node_def = minetest.registered_nodes[node.name]
+			if node_def then
+				-- Buckets will run a node's on_punch function if it is not liquid.
+				if node_def.on_punch then
+					giving_back = node_def.on_punch(
```

Comment:
You also can't guarantee that the on_punch is defined in the node definition. If it's not this might throw an error.

You should use [minetest.node_punch](http://dev.minetest.net/minetest.node_punch).